src/hotspot/cpu/aarch64/gc/z/zGlobals_aarch64.cpp

```diff
@@ -135,9 +138,66 @@
 //  * 63-48 Fixed (16-bits, always zero)
 //
 
+// Default value if probing is not implemented for a certain platform: 128TB
+#define DEFAULT_MAX_ADDRESS_BIT 47
+// Minimum value returned, if probing fails: 64GB
+#define MINIMUM_MAX_ADDRESS_BIT 36
+
+static size_t probe_valid_max_address_bit() {
+#ifdef LINUX
+  size_t max_address_bit = 0;
+  const size_t page_size = os::vm_page_size();
+  for (int i = DEFAULT_MAX_ADDRESS_BIT; i > MINIMUM_MAX_ADDRESS_BIT && max_address_bit == 0; --i) {
+    const uintptr_t base_addr = ((uintptr_t) 1U) << i;
+    if (msync((void*)base_addr, page_size, MS_ASYNC) == 0) {
+      // msync suceeded, the address is valid, and maybe even already mapped.
+      max_address_bit = i;
+      break;
+    }
+    if (errno != ENOMEM) {
+      // Some error occured. This should never happen, but msync
+      // has some undefined behavior, hence ignore this bit.
+#ifdef ASSERT
+      fatal("Received %s while probing the address space for the highest valid bit", os::errno_name(errno));
+#else // ASSERT
+      log_warning(gc)("Received %s while probing the address space for the highest valid bit", os::errno_name(errno));
+#endif // ASSERT
+      continue;
+    }
+    // Since msync failed with ENOMEM, the page might not be mapped.
+    // Try to map it, to see if the address is valid.
+    void* result_addr = mmap((void*) base_addr, page_size, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0);
+    if ((uintptr_t) result_addr == base_addr) {
+      // address is valid
+      max_address_bit = i;
+    }
+    if (result_addr != MAP_FAILED) {
+      munmap(result_addr, page_size);
+    }
+  }
```

Comment:
If you swap the order of these if-statements you could add a break after 172.

This way the exit of the loop would look like the one you have after msync. You would also be able to get rid of the ' && max_address_bit == 0' check in the for statement.